Iterators

manuelma opened the pull request:
Some optimisations:
- Use iterators for constraint and variable indices rather than vectors.
- Don't traverse iterators given to `active_stochastic_paths()` unless there's more than one `stochastic_scenario` in the model.
- Remove `units_available` and keep only `units_on`.
- Don't build the auto-updating structure if the model is not rolling.
- Don't generate unit commitment variables unless they are needed.
- Use variable bounds instead of constraints whenever possible.
- `use_connection_intact_flow` is false by default.
- Introduce `use_tight_compact_formulations`, false by default.

Re https://github.com/spine-tools/SpineOpt.jl/issues/855

## Checklist before merging
- [ ] Documentation is up-to-date
- [ ] Unit tests have been added/updated accordingly
- [ ] Code has been formatted according to SpineOpt's style
- [ ] Unit tests pass

manuelma committed:
Use iterators for indices

manuelma committed:
Remove unique in constraint indices

manuelma committed:
Speed up variable generation

manuelma committed:
Speed up some variable indices and constraints using get

manuelma committed:
Fix constraint unit capacity margin indices

manuelma committed:
Merge branch 'master' into iterators

manuelma committed:
Introduce use_tight_compact_formulations

manuelma committed:
Merge remote-tracking branch 'origin/master' into iterators

manuelma committed:
Improve connection flow capacity constraint

manuelma committed:
Remove unnecessary start up / shut down variables

manuelma committed:
Don't build auto-updating model stuff if not rolling

gnawin commented:
@manuelma, I've experienced the same errors for Julia 1.6 and only for 1.6 in PR #909. Do you happen to know how to fix it? Thanks.

manuelma committed:
Replace flow and state capacity constraints with variable bounds

...whenever possible

manuelma committed:
Fix tests

manuelma committed:
Skip connection_flow_capacity_bidirectional if !use_tight_compact_formulations

manuelma committed:
Fix tests

manuelma committed:
Make sure stage models auto-update with benders

If the subproblem has only 1 window it still needs to be able to
rewind, so it needs the auto-update structure.

manuelma committed:
Merge remote-tracking branch 'origin/iterators' into iterators

manuelma committed:
Merge branch 'master' into iterators

manuelma committed:
Merge branch 'master' into iterators

manuelma commented:
I am splitting this into more self-contained PRs - one for each optimisation. Closing now.
